### ISSUE TITLE
Gradle 8.5 Memory leak when building from Android Studio

### ISSUE THREAD

**Issue by eygraber** [gradle/gradle#27491]
### Current Behavior

When I build in Android Studio, the Gradle daemon process takes a large amount of memory, very quickly going up to my `Xmx` value and beyond on the first build (12gb+). When building from the command line this doesn't happen (8gb after ~5 builds).

### Expected Behavior

This much memory shouldn't be getting used.

### Context (optional)

I filed an issue on Android Studio that seems similar because their RAM usage goes through roof (no OOM and eventually eats through all of my RAM until the OOM killer kicks in) - https://issuetracker.google.com/issues/316783745

I don't have a repro project, and can't share the project this is happening on. I did capture a thread and heap dump, as well as some screenshots from flight control.

The file is too big to share through GitHub (3.2gb). What's the best way I can share that?

### Steps to Reproduce

Build my project through Android Studio.

### Gradle version

8.5

**Comment by ov7a** [comment]
Sorry that you're having trouble with Gradle!

This is an issue with Android Studio Gradle integration, which is developed by Google and cannot be addressed in the Gradle Build Tool project.

***

Unless the report on Google's issue tracker points to Gradle-specific problem, we believe this is not something that can be addressed by us, thus we are closing this issue. Please comment if you have additional information on that.

**Comment by eygraber** [comment]
Is this a known issue? Otherwise I think it's worth taking a look at the thread and heap dump from gradle just to make sure it's not something in gradle.

**Comment by ov7a** [comment]
@eygraber Sorry, our resources are too limited to do that. If you think it's a Gradle issue, please provide an isolated reproducer.

The ideal way to provide a reproducer is to leverage our [reproducer template](https://github.com/gradle/gradle-issue-reproducer). You can also use [Gradle Project Replicator](https://github.com/android/project-replicator) to reproduce the structure of your project.

**Comment by eygraber** [comment]
I update AS to the newest canary, and now the AS RAM usage is fine, but Gradle's has shot through the roof, which makes me think this is really a Gradle issue.

I can try using the reproducer, but this is a complex project, and without knowing what is causing the leak it would be hard to reproduce. I have a thread dump and heap dump, and I think someone from Gradle would be better suited to look at it than I am.

**Comment by ritesh-singh** [comment]
I noticed similar issue.

After upgrading AGP 8.2.0 ,  Gradle >= 8.3 Kotlin 1.9.21 and other dependent kotlin compiler libraries like KSP, Compose compiler etc... while running AGP link task.
PermGen quickly take ups the whole heap memory - while gradle task is still in configuration phase.
After analysing the heap dump, looks like LintModelWriterTask from AGP(8.2.0) lint and DefaultUnlockedConfiguration from Gradle(8.5) are major contributors to OOM.


LintModelWriterTask is held by VisitableURLClassLoader, which seems to be coming from Gradle 8.5
Tried downgrading the Gradle version to 8.4 and 8.3, but no luck!
